### PR TITLE
bpo-41810: Reintroduce `types.EllipsisType`, `.NoneType` & `.NotImplementedType`

### PR DESCRIPTION
**Pull request by BvB93**
See the relevant BPO for more details: https://bugs.python.org/issue41810.

https://bugs.python.org/issue41810

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@BvB93

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by BvB93** [comment]
For reference: I just signed the CLA a few minutes ago.

File: Misc/NEWS.d/next/Library/2020-09-20-15-14-05.bpo-41810.7l8lyV.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+:data:`types.EllipsisType`, :data:`types.EllipsisType` and
+:data:`types.NoneType` have been reintroduced.
```
Comment (ericvsmith) [inline Misc/NEWS.d/next/Library/2020-09-20-15-14-05.bpo-41810.7l8lyV.rst]: We should mention the reason behind this reintroduction: for type checkers which can't use `type(Ellipsis)`, etc.

File: Misc/NEWS.d/next/Library/2020-09-20-15-14-05.bpo-41810.7l8lyV.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+:data:`types.EllipsisType`, :data:`types.EllipsisType` and
```
Comment (ericvsmith) [inline Misc/NEWS.d/next/Library/2020-09-20-15-14-05.bpo-41810.7l8lyV.rst]: One of these should be `type.NotImplementedType`.

**Review by ericvsmith — CHANGES_REQUESTED** [review]
I've not done a thorough review yet. These comments are just about the blurb.

**Comment by gvanrossum** [comment]
The PR would be easier to review if you hadn’t tried to fix all occurrences of type(......). That could then be done in one or several follow up PRs. Or possibly just skipped until there is demand.

File: Misc/NEWS.d/next/Library/2020-09-20-15-14-05.bpo-41810.7l8lyV.rst (hunk 2 above)
Comment (BvB93) [reply Misc/NEWS.d/next/Library/2020-09-20-15-14-05.bpo-41810.7l8lyV.rst]: Good catch; fixed in https://github.com/python/cpython/pull/22336/commits/f682623aa3001a4ac19e727c3103cdb0dce8672e.

File: Misc/NEWS.d/next/Library/2020-09-20-15-14-05.bpo-41810.7l8lyV.rst (hunk 1 above)
Comment (BvB93) [reply Misc/NEWS.d/next/Library/2020-09-20-15-14-05.bpo-41810.7l8lyV.rst]: I added a few words about the motivation in https://github.com/python/cpython/pull/22336/commits/370149882d2cea95f2fed7c732bcbb6ae343a3fe.

File: Misc/NEWS.d/next/Library/2020-09-20-15-14-05.bpo-41810.7l8lyV.rst [hunk 3]
```diff
@@ -1,2 +1,3 @@
-:data:`types.EllipsisType`, :data:`types.EllipsisType` and
-:data:`types.NoneType` have been reintroduced.
+:data:`types.EllipsisType`, :data:`types.NotImplementedType` and
+:data:`types.NoneType` have been reintroduced, aforementioned types
+now being useable as annotations.
```
Comment (ericvsmith) [inline Misc/NEWS.d/next/Library/2020-09-20-15-14-05.bpo-41810.7l8lyV.rst]: I'm not trying to be overly picky here, but I really think you need to mention type checkers. If I was reading this comment 5 years from now, I'd think "but type(Ellipsis) was already usable as an annotation, so we should delete EllipsisType".

**Comment by BvB93** [comment]
> The PR would be easier to review if you hadn’t tried to fix all occurrences of type(......). That could then be done in one or several follow up PRs. Or possibly just skipped until there is demand.

If you want I can remove those changes; they're localized to https://github.com/python/cpython/pull/22336/commits/0a97a885178adaf79d93041aed5bbab3c9e7d264 and https://github.com/python/cpython/pull/22336/commits/2791a239bacfe1f16e9672ca063cdb1f29e9a147 so reverting them should be straightforward.

**Comment by gvanrossum** [comment]
> If you want I can remove those changes; they're localized to 0a97a88 and 2791a23 so reverting them should be straightforward.

@ericvsmith, what do you think?

File: Misc/NEWS.d/next/Library/2020-09-20-15-14-05.bpo-41810.7l8lyV.rst (hunk 3 above)
Comment (BvB93) [reply Misc/NEWS.d/next/Library/2020-09-20-15-14-05.bpo-41810.7l8lyV.rst]: Fair enough, fixed in https://github.com/python/cpython/pull/22336/commits/33caa6fb32ec98766a9402670176ddde16d47738.

**Comment by ericvsmith** [comment]
I do think it would be easier to review if they were moved to a separate PR, even if it's on the same bpo.

**Comment by BvB93** [comment]
> I do think it would be easier to review if they were moved to a separate PR, even if it's on the same bpo.

The changes have been reverted as of https://github.com/python/cpython/pull/22336/commits/84ebe183e0c4ed9059e88fbede11092759f3eb66.

**Review by gvanrossum — COMMENTED** [review]
I think the docs for the three special objects should not start with “The sole value of the type ...”.

The first sentence should serve as a brief summary of what the object is used for. You can then follow up with the detail about its singleton type.

Otherwise LGTM. Thanks for your patience!

**Review by ericvsmith — APPROVED** [review]
This looks good to me!

**Comment by ericvsmith** [comment]
> I think the docs for the three special objects should not start with “The sole value of the type ...”.

It's just mirroring the existing documentation for `None`. But I agree that moving the "the sole value" sentence (or sentence fragment) to the end of the description would be an improvement.

File: Doc/library/constants.rst [hunk 4]
```diff
@@ -19,14 +19,15 @@ A small number of constants live in the built-in namespace.  They are:
 
 .. data:: None
 
-   The sole value of the type ``NoneType``.  ``None`` is frequently used to
+   The sole value of the type :data:`types.NoneType`. ``None`` is frequently used to
```
Comment (terryjreedy) [inline Doc/library/constants.rst]: "The sole value of the type NoneType" is most easily read (parsed) as referring to NoneType's usefulness.  I expected it to be followed by 'is to ...' and initially read 'is frequently' as a grammatical error.  It took awhile to realize that this is a convoluted  way to say 'None'.  I strongly recommend a rewording.

```suggestion
   ''None'' is the only instance of :data:`types.NoneType`.  It is frequently used to
```
`types.NoneType` is already twice identified as a type, so a third mention is not needed.

File: Doc/library/constants.rst [hunk 5]
```diff
@@ -19,14 +19,15 @@ A small number of constants live in the built-in namespace.  They are:
...
    represent the absence of a value, as when default arguments are not passed to a
    function. Assignments to ``None`` are illegal and raise a :exc:`SyntaxError`.
 
 
 .. data:: NotImplemented
 
-   Special value which should be returned by the binary special methods
+   Sole value of the type :data:`types.NotImplementedType` and a
+   special value which should be returned by the binary special methods
```
Comment (terryjreedy) [inline Doc/library/constants.rst]: See above.  Use a parallel construction.

```suggestion
   ``NotImplemented is the only instance of :data:`types.NotImplementedType`.  It is a
   special value which should be returned by the binary special methods
```

File: Doc/library/constants.rst [hunk 6]
```diff
@@ -59,7 +60,8 @@ A small number of constants live in the built-in namespace.  They are:
 .. index:: single: ...; ellipsis literal
 .. data:: Ellipsis
 
-   The same as the ellipsis literal "``...``".  Special value used mostly in conjunction
+   The same as the ellipsis literal "``...``" and sole value of the type
```
Comment (terryjreedy) [inline Doc/library/constants.rst]:
```suggestion
   The same as the ellipsis literal "``...``" and the only instance of
```

File: Doc/whatsnew/3.10.rst [hunk 7]
```diff
@@ -145,6 +145,14 @@ Add :data:`sys.orig_argv` attribute: the list of the original command line
 arguments passed to the Python executable.
 (Contributed by Victor Stinner in :issue:`23427`.)
 
+types
+-----
+
+Reintroduced the :data:`types.EllipsisType`, :data:`types.NoneType`
+and :data:`types.NotImplementedType` classes, providing a new set
+of types readily interpretable by type checkers.
```
Comment (terryjreedy) [inline Doc/whatsnew/3.10.rst:153]: I think this and the blurb are fine, providing a reason without saying either that this is the only reason or that these should be use everywhere.

**Comment by terryjreedy** [comment]
I did not particularly want the change in IDLE, because 1) I am not sure I prefer it (partly because of the extra import) and I don't know that PEP-8 'mandates' it*, and 2) I  cannot backport it to actively maintained 3.8 and 3.9 code, so it would cause a merge conflict if I patched that section of code.  I checked that the latter seems unlikely in this particular case.

I think the effort to 'sanitize' all existing code is futile, since people *will* continue to put `type(None)` in code.   Consider leading existing code alone.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
Approved, but only with misleading wording fixed.

File: Doc/library/constants.rst (hunk 4 above)
Comment (BvB93) [reply Doc/library/constants.rst]: Based on your and Guido's feedback (https://github.com/python/cpython/pull/22336#pullrequestreview-493014262) I'm proposing something like this:

```
.. data:: None

   An object frequently used to represent the absence of a value, as when
   default arguments are not passed to a function. Assignments to ``None``
   are illegal and raise a :exc:`SyntaxError`.
   Sole instance of the :data:`types.NoneType` type.
```
Any thoughts?

Comment (ericvsmith) [reply Doc/library/constants.rst]: That's much better, I think. Although I'd make the last sentence (with the correct markup): "None is the sole instance of the NoneType type."

**Comment by BvB93** [comment]
@terryjreedy the documentation changes have been implemented as of https://github.com/python/cpython/pull/22336/commits/23ed559738432aae3e2cca47636c75f15abc341a.

**Review by terryjreedy — APPROVED** [review]

**Review by gvanrossum — APPROVED** [review]

**Pull request merged**